File: benchmark_client/scripts/cpp/Zombie Apocalypse.cpp
#include <algorithm>
#include <chrono>
#include <cstdio>
#include <random>
#include <vector>

enum { DEAD = 0, HUMAN = 1, ZOMBIE = 2 };

struct Agent {
    int id;
    int type;
    int x;
    int y;
};

static inline int cas(std::vector<int> &grid, int addr, int cmp, int val) {
    const int old = grid[static_cast<std::size_t>(addr)];
    if (old == cmp) {
        grid[static_cast<std::size_t>(addr)] = val;
    }
    return old;
}

static void simulate_step(std::vector<Agent> &agents, std::vector<int> &grid, int nx, int ny,
                          std::mt19937 &gen) {
    std::uniform_int_distribution<int> dist_dir(0, 4);
    std::uniform_real_distribution<float> dist_u01(0.0f, 1.0f);
    constexpr float p_win = 0.1f;

    const int na = static_cast<int>(agents.size());
    std::vector<int> order(static_cast<std::size_t>(na));
    for (int i = 0; i < na; ++i) {
        order[static_cast<std::size_t>(i)] = i;
    }
    std::shuffle(order.begin(), order.end(), gen);

    for (int k = 0; k < na; ++k) {
        const int idx = order[static_cast<std::size_t>(k)];
        if (agents[static_cast<std::size_t>(idx)].type == DEAD) {
            continue;
        }
        const int ax = agents[static_cast<std::size_t>(idx)].x;
        const int ay = agents[static_cast<std::size_t>(idx)].y;
        int nx2 = ax;
        int ny2 = ay;
        const int d = dist_dir(gen);
        if (d == 1) {
            ny2 = (ay - 1 + ny) % ny;
        } else if (d == 2) {
            ny2 = (ay + 1) % ny;
        } else if (d == 3) {
            nx2 = (ax - 1 + nx) % nx;
        } else if (d == 4) {
            nx2 = (ax + 1) % nx;
        }
        const int target_idx = ny2 * nx + nx2;
        const int old_cell = ay * nx + ax;
        const int old = cas(grid, target_idx, 0, idx + 1);
        if (old == 0 || old == idx + 1) {
            if (d != 0 && old == 0) {
                cas(grid, old_cell, idx + 1, 0);
            }
            agents[static_cast<std::size_t>(idx)].x = nx2;
            agents[static_cast<std::size_t>(idx)].y = ny2;
        } else {
            cas(grid, old_cell, 0, idx + 1);
        }
    }

    std::shuffle(order.begin(), order.end(), gen);
    for (int k = 0; k < na; ++k) {
        const int idx = order[static_cast<std::size_t>(k)];
        if (agents[static_cast<std::size_t>(idx)].type != HUMAN) {
            continue;
        }
        const int mx = agents[static_cast<std::size_t>(idx)].x;
        const int my = agents[static_cast<std::size_t>(idx)].y;
        const int nbr[4] = {
            ((my - 1 + ny) % ny) * nx + mx,
            ((my + 1) % ny) * nx + mx,
            my * nx + ((mx - 1 + nx) % nx),
            my * nx + ((mx + 1) % nx),
        };
        for (int gid : nbr) {
            const int nid = grid[static_cast<std::size_t>(gid)] - 1;
            if (nid >= 0 && agents[static_cast<std::size_t>(nid)].type == ZOMBIE) {
                if (dist_u01(gen) < p_win) {
                    agents[static_cast<std::size_t>(nid)].type = DEAD;
                    grid[static_cast<std::size_t>(gid)] = 0;
                } else {
                    agents[static_cast<std::size_t>(idx)].type = ZOMBIE;
                }
                break;
            }
        }
    }
}

int main() {
    constexpr int NX = 150;
    constexpr int NY = 150;
    constexpr int NL = 1500;
    constexpr int NZ = 50;
    constexpr int STEPS = 500;
    const int total = NL + NZ;
    const int grid_size = NX * NY;

    std::vector<Agent> agents(static_cast<std::size_t>(total));
    std::vector<int> grid(static_cast<std::size_t>(grid_size), 0);
    std::vector<int> positions(static_cast<std::size_t>(grid_size));
    for (int i = 0; i < grid_size; ++i) {
        positions[static_cast<std::size_t>(i)] = i;
    }

    std::mt19937 gen(12345u);
    std::shuffle(positions.begin(), positions.end(), gen);

    for (int i = 0; i < total; ++i) {
        const int p = positions[static_cast<std::size_t>(i)];
        agents[static_cast<std::size_t>(i)].id = i;
        agents[static_cast<std::size_t>(i)].type = (i < NL) ? HUMAN : ZOMBIE;
        agents[static_cast<std::size_t>(i)].x = p % NX;
        agents[static_cast<std::size_t>(i)].y = p / NX;
        grid[static_cast<std::size_t>(agents[static_cast<std::size_t>(i)].y * NX +
                                      agents[static_cast<std::size_t>(i)].x)] = i + 1;
    }

    const auto t0 = std::chrono::high_resolution_clock::now();
    for (int t = 0; t < STEPS; ++t) {
        simulate_step(agents, grid, NX, NY, gen);
    }
    const auto t1 = std::chrono::high_resolution_clock::now();
    const double elapsed_ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

    (void)agents[0].type;
    std::printf("{\"execution_time_ms\": %.6f}\n", elapsed_ms);
    return 0;
}

File: benchmark_client/scripts/go/Zombie Apocalypse.go
package main

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"time"
)

const (
	dead      = 0
	human     = 1
	zombie    = 2
	nx        = 150
	ny        = 150
	nHumans   = 1500
	nZombies  = 50
	steps     = 500
	pWinHuman = 0.1
)

type agent struct {
	typ int
	x   int
	y   int
}

func cas(grid []int32, addr, cmp, val int) int32 {
	old := grid[addr]
	if old == int32(cmp) {
		grid[addr] = int32(val)
	}
	return old
}

func simulateStep(agents []agent, grid []int32, src *rand.Rand) {
	na := len(agents)
	order := make([]int, na)
	for i := range order {
		order[i] = i
	}
	src.Shuffle(na, func(i, j int) { order[i], order[j] = order[j], order[i] })

	for _, ii := range order {
		idx := ii
		if agents[idx].typ == dead {
			continue
		}
		ax := agents[idx].x
		ay := agents[idx].y
		nxp := ax
		nyp := ay
		d := src.Intn(5)
		switch d {
		case 1:
			nyp = (ay - 1 + ny) % ny
		case 2:
			nyp = (ay + 1) % ny
		case 3:
			nxp = (ax - 1 + nx) % nx
		case 4:
			nxp = (ax + 1) % nx
		}
		target := nyp*nx + nxp
		oldCell := ay*nx + ax
		old := cas(grid, target, 0, idx+1)
		if old == 0 || old == int32(idx+1) {
			if d != 0 && old == 0 {
				cas(grid, oldCell, idx+1, 0)
			}
			agents[idx].x = nxp
			agents[idx].y = nyp
		} else {
			cas(grid, oldCell, 0, idx+1)
		}
	}

	src.Shuffle(na, func(i, j int) { order[i], order[j] = order[j], order[i] })
	for _, ii := range order {
		idx := ii
		if agents[idx].typ != human {
			continue
		}
		mx := agents[idx].x
		my := agents[idx].y
		nbr := []int{
			((my-1+ny)%ny)*nx + mx,
			((my+1)%ny)*nx + mx,
			my*nx + ((mx-1+nx)%nx),
			my*nx + ((mx+1)%nx),
		}
		for _, gid := range nbr {
			nid := int(grid[gid]) - 1
			if nid >= 0 && agents[nid].typ == zombie {
				if src.Float64() < pWinHuman {
					agents[nid].typ = dead
					grid[gid] = 0
				} else {
					agents[idx].typ = zombie
				}
				break
			}
		}
	}
}

func main() {
	total := nHumans + nZombies
	agents := make([]agent, total)
	grid := make([]int32, nx*ny)
	src := rand.New(rand.NewSource(12345))
	pos := src.Perm(nx * ny)

	for i := 0; i < total; i++ {
		agents[i].typ = human
		if i >= nHumans {
			agents[i].typ = zombie
		}
		p := pos[i]
		agents[i].x = p % nx
		agents[i].y = p / nx
		grid[agents[i].y*nx+agents[i].x] = int32(i + 1)
	}

	start := time.Now()
	for t := 0; t < steps; t++ {
		simulateStep(agents, grid, src)
	}
	elapsedMs := float64(time.Since(start).Nanoseconds()) / 1e6

	_ = agents[0].typ
	out, err := json.Marshal(map[string]float64{"execution_time_ms": elapsedMs})
	if err != nil {
		os.Exit(1)
	}
	fmt.Println(string(out))
}
